packages/fhir-router/src/batch.ts

```diff
@@ -191,20 +192,27 @@ class BatchProcessor {
   }
 
   private rewriteIdsInArray(input: any[]): any[] {
-    return input.map((item) => this.rewriteIds(item));
+    return input.map((item, index) => this.rewriteIds(item, index));
   }
 
   private rewriteIdsInObject(input: any): any {
-    return Object.fromEntries(Object.entries(input).map(([k, v]) => [k, this.rewriteIds(v)]));
+    return Object.fromEntries(Object.entries(input).map(([k, v]) => [k, this.rewriteIds(v, k)]));
   }
 
-  private rewriteIdsInString(input: string): string {
+  private rewriteIdsInString(input: string, key?: string | number): string {
     const matches = input.match(/urn:uuid:\w{8}-\w{4}-\w{4}-\w{4}-\w{12}/);
     if (matches) {
       const fullUrl = matches[0];
       const resource = this.ids[fullUrl];
       if (resource) {
-        return input.replaceAll(fullUrl, getReferenceString(resource));
+        let referenceString: string | undefined = getReferenceString(resource);
+        // If this is the top-level 'id' field of the resource, we want to remove the `resourceType` prefix from the
+        // reference string
+        if (key === 'id') {
```

Comment:
This will rewrite `id` in nested objects too, yeah?

Probably not a big deal, because highly unlikely to have the `urn:uuid:` syntax in a nested `id` property.

If we care, we could have an outer function that checks `id` at the top level.

Reply:
Yeah it will. I was ignoring that for now, but the right thing is probably just to do a top-level check. I'll make the fix